File: p39_permutation.cpp
#include <iostream>
#include <algorithm>

const int MAX_N = 1000000;

bool used[MAX_N];
int perm[MAX_N];

int N;
// get n! pattern

void permutation1(int pos, int n){
	if (pos == n){

		// processing perm
		for(int j=0; j<n; j++) std::cout << perm[j];
		std::cout << '\n';
		return ;
	}

	// loop to pick up pos of perm from 0...n-1
	for (int i=0; i<n; i++){
		if(!used[i]){
			perm[pos] = i; 
			used[i] = true;
			permutation1(pos + 1, n);
			used[i] = false;
		} 
	} 
	return ;
}

int perm2[MAX_N];

void permutation2(int n){
	for(int i=0; i<n; i++){
		perm2[i] = i;
	}
	do {
		// process perm2
		for(int j=0; j<n; j++) std::cout << perm2[j];
		std::cout << '\n';
	} while (std::next_permutation(perm2, perm2+n));

	return ;
}

int main(){

	std::cout << "Input N :";
	std::cin >> N;
	for(int i=0; i<N; i++) used[i]=false;
	std::cout << "------------------Recursive Function----------------" << '\n';
	permutation1(0,N);
	std::cout << "------------------------Library---------------------" << '\n';
	permutation2(N);

}

File: p39_permutation.py
import itertools

n = int( input("Input N:") )

used = [False for i in range(n)]
perm = [0 for i in range(n)]

def permutation(pos, n):

    if pos == n:
        print("".join( [ str(elem) for elem in perm ] ))

    for i in range(n):
        if not used[i]:
            perm[pos] = i
            used[i] = True
            permutation(pos+1,n)
            used[i] = False

print("-----Recursive Function-----")
permutation(0,n)

print("----------Library-----------")
for perm in itertools.permutations(range(n),n):
    print( "".join( [str(elem) for elem in perm] ) )